perllib/FixMyStreet/Cobrand/Bexley/Waste.pm

```diff
@@ -51,23 +58,71 @@ sub bin_services_for_address {
     };
 
     my $site_services = $self->whitespace->GetSiteCollections($property->{uprn});
-
-    # TODO: Filter out services with SiteServiceValidTo in the past or SiteServiceValidFrom in the future
-    $site_services = [ grep { $_->{NextCollectionDate} } @$site_services ];
-
-    my $services = [ map {
-        {
-            id => $_->{SiteServiceID},
-            service_name => $_->{ServiceItemDescription},
-            next => {
-                date => $_->{NextCollectionDate},
-                ordinal => 'th', # TODO!
-                changed => 0,
-            },
+    my @site_services_filtered;
+
+    my $now_dt = DateTime->now->set_time_zone( FixMyStreet->local_time_zone );
+    for my $service (@$site_services) {
+        # TODO Do we need to filter out services where NextCollectionDate is
+        # in the past?
+        next if !$service->{NextCollectionDate};
+
+        my $from_dt = eval {
+            DateTime::Format::W3CDTF->parse_datetime(
+                $service->{SiteServiceValidFrom} );
+        };
+        if ($@) {
+            warn $@;
+            next;
         }
-    } @$site_services ];
+        next if $now_dt < $from_dt;
+
+        # 0001-01-01T00:00:00 seems to represent an undefined date
+        if ( $service->{SiteServiceValidTo} ne '0001-01-01T00:00:00' ) {
+            my $to_dt = eval {
+                DateTime::Format::W3CDTF->parse_datetime(
+                    $service->{SiteServiceValidTo} );
+            };
+            if ($@) {
+                warn $@;
+                next;
+            }
+            next if $now_dt > $to_dt;
+        }
+
+        push @site_services_filtered, $service;
+    }
+
+    my $services = [
+        map {
+            my $next_dt = eval {
+                DateTime::Format::W3CDTF->parse_datetime(
+                    $_->{NextCollectionDate} );
+            };
+            if ($@) {
+                warn $@;
+                next;
+            }
```

Comment:
This `next` won't work as expected inside the map call.

You can see the problem by adding an extra service in the tests (inside `_site_collections`). If you create a service 7 by copying service 6 and give it an invalid `NextCollectionDate` then you should get an error about "Exiting subroutine via next".

Think the fix is to ignore services with an invalid `NextCollectionDate` before this map call.